framework/resource/metricsresource/resource.go

```diff
@@ -181,13 +180,5 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateState inter
 }
 
 func (r *Resource) Underlying() framework.Resource {
-	underlying := r.resource
-	for {
-		wrapper, ok := underlying.(internal.Wrapper)
-		if ok {
-			underlying = wrapper.Underlying()
-		} else {
-			return underlying
-		}
-	}
+	return r.resource
```

Comment:
I do not understand the wrapping magic here. IMO anything implementing `Underlying() framework.Resource` should either return itself or the direct underlying resource. Not more. The next resource will do the same and at some point the last underlying will return itself and we are good. This code here seemed to never return. It was just looping forever which made operators not reconcile anymore.

Reply:
Oh, yeah, there is infinite loop when the underlying resource implements Underlying itself. I didn't anticipate that case.

At the state before  you merged this PR our Resources shouldn't have `Underlying() Resource` method. If you would just remove `Underlying` methods from all resources the thing should work.

I am sorry you had to deal with that. I should have update one of the operators and check that.